src/main/java/org/truffleruby/language/globals/ReadMatchReferenceNodes.java

```diff
@@ -0,0 +1,315 @@
+
+        protected final ConditionProfile matchNilProfile = ConditionProfile.createBinaryProfile();
+
+        public ReadPostMatchNode(GetFromThreadAndFrameLocalStorageNode readMatchNode) {
+            this.readMatchNode = readMatchNode;
+        }
+
+        @Override
+        public Object execute(VirtualFrame frame) {
+            final Object match = readMatchNode.execute(frame);
+
+            if (matchNilProfile.profile(match == nil())) {
+                return nil();
+            }
+
+            final DynamicObject matchData = (DynamicObject) match;
+
+            return postMatchNode.execute(matchData);
+        }
+
+        @Override
+        public Object isDefined(VirtualFrame frame) {
+            if (isNil(execute(frame))) {
+                return nil();
+            } else {
+                return coreStrings().GLOBAL_VARIABLE.createInstance();
+            }
+        }
+    }
+
+    public static class ReadMatchNode extends RubyNode {
+        @Child MatchNode tosMatchNode = MatchNode.create();
+        @Child private GetFromThreadAndFrameLocalStorageNode readMatchNode;
+
+        protected final ConditionProfile matchNilProfile = ConditionProfile.createBinaryProfile();
+
+        public ReadMatchNode(GetFromThreadAndFrameLocalStorageNode readMatchNode) {
+            this.readMatchNode = readMatchNode;
+        }
+
+        @Override
+        public Object execute(VirtualFrame frame) {
+            final Object match = readMatchNode.execute(frame);
+
+            if (matchNilProfile.profile(match == nil())) {
+                return nil();
+            }
+
+            final DynamicObject matchData = (DynamicObject) match;
+
+            return tosMatchNode.execute(matchData);
+        }
+
+        @Override
+        public Object isDefined(VirtualFrame frame) {
+            if (isNil(execute(frame))) {
+                return nil();
+            } else {
+                return coreStrings().GLOBAL_VARIABLE.createInstance();
+            }
+        }
+    }
+
+    public static class ReadHighestMatchNode extends RubyNode {
+        @Child private GetFromThreadAndFrameLocalStorageNode readMatchNode;
+        @Child private ValuesNode getValues = ValuesNode.create();
+
+        protected final ConditionProfile matchNilProfile = ConditionProfile.createBinaryProfile();
+
+        public ReadHighestMatchNode(GetFromThreadAndFrameLocalStorageNode readMatchNode) {
+            this.readMatchNode = readMatchNode;
+        }
+
+        @Override
+        public Object execute(VirtualFrame frame) {
+            final Object match = readMatchNode.execute(frame);
+
+            if (matchNilProfile.profile(match == nil())) {
+                return nil();
+            }
+
+            final DynamicObject matchData = (DynamicObject) match;
+
+            final Object[] values = getValues.execute(matchData);
+
+            for (int n = values.length - 1; n >= 0; n--)
```

Comment:
Just my style preference, I think it's nicer to have the `{ }` here

Reply:
It was part of the old code that I didn't change, Git is showing it as a new file but it's really a rename and some re-org.

Reply:
I think it's generally good to do a bit of code cleanup when moving things around, at least basic stuff like this.